Beginning of new Sphinx tutorial

astrojuanlu opened the pull request:
### Feature or Bugfix
- Feature

### Purpose
This is the first pull request (hopefully out of many more to come) for the new Sphinx tutorial proposed in https://github.com/sphinx-doc/sphinx/issues/9165. Paraphrasing @evildmp, it is [by no means finished, but almost complete](https://www.writethedocs.org/videos/portland/2021/always-complete-never-finished-daniele-procida/) 🙂 

Rendered version: https://sphinx--9276.org.readthedocs.build/en/9276/tutorial/index.html

Since this is my first substantial contribution to the Sphinx project, I wanted to open this small pull request early to address any potential disagreements the Sphinx maintainers might have about things like the tone, the structure, the link on the front page, or anything else.

This tutorial partially covers sections 2-4 of the table of contents originally proposed in #9165:

2. About this tutorial
3. Getting started
4. First steps to document our project using Sphinx

There is room for those sections to grow in subsequent pull requests, they are not cast in stone. For now, I wanted to set the overall intent and structure of the whole tutorial. As a result, I might slightly change or reword these contents in the near future, to accommodate for its growth.

Some noteworthy decisions reflected implicitly or explicitly in the discussion in #9165:

- Inclusion of the Tutorial in the front page
- Tutorial "aimed towards Sphinx newcomers"
- Assumes Linux-like command line. With some work, we could use tabs for different operative systems [like the Django documentation does](https://docs.djangoproject.com/en/dev/intro/tutorial01/#writing-your-first-django-app-part-1), but I didn't want to overcomplicate things just yet.
  - Still, we might want to offer a hint to macOS users (to reassure them that their terminal is fine) and to Windows users (to encourage them to use WSL2?)
- Explicit usage of Python virtual environments, instead of other installation methods.
- Creation of a fictional library. I think this makes the tutorial a bit more playful and interesting, instead of using filler/dummy text all the time. I took this idea from [this Sphinx tutorial](https://brandons-sphinx-tutorial.readthedocs.io/) by @brandon-rhodes (and left some Easter eggs for the attentive reader)

Looking forward to hearing Sphinx maintainers' thoughts on this!

cc @ericholscher 

### Relates
- #9165

astrojuanlu committed:
Tutorial link in front page, audience, objectives, and prerequisites

astrojuanlu committed:
Beginning of the tutorial, tweaks to introduction

astrojuanlu commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,152 @@
+with the objective of properly documenting it.
+
+To showcase Sphinx automatic documentation generation capabilities
+we will use Python, which is the default :term:`domain`
```
```suggestion
you will use Python, which is the default :term:`domain`
```

astrojuanlu commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,152 @@
+To follow the instructions you will need access to a Linux-like command line
+and a basic understanding of how it works,
+as well as a working Python installation for development,
+since we will use *Python virtual environments* to create our project
```
```suggestion
since you will use *Python virtual environments* to create the project
```

astrojuanlu commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,152 @@
+   that creates recipes mixing random ingredients.
+
+It is a good moment to create a Python virtual environment
+and install the tools we will need.
```
```suggestion
and install the required tools.
```

astrojuanlu commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,152 @@
+  and contains the root of the "table of contents tree" (or *toctree*).
+
+Thanks to this bootstrapping step,
+we already have everything we need
```
```suggestion
you already have everything needed
```

astrojuanlu commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,152 @@
+
+Thanks to this bootstrapping step,
+we already have everything we need
+to render our documentation as HTML for the first time.
```
```suggestion
to render the documentation as HTML for the first time.
```

astrojuanlu committed:
Style

astrojuanlu committed:
Add index tweaks

tk0miya commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,198 @@
+This showcases several features of the reStructuredText syntax, including:
+
+- a **section header** using ``===`` for the underline,
+- two examples of **inline markup**: ``**bold**`` and ``*italics*``,
```
FYI: Accurately, these are `**strong emphasis**` and `*emphasis*`. Surely they're rendered as bold and italic usually. I don't know which wording is better for beginners. But we need to know they're different things and choose better representation.
https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#emphasis

refs: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/strong#%3Cb%3E_vs._%3Cstrong%3E

astrojuanlu replied:
You're right, let's use the original terminology here. Changing!

tk0miya commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,198 @@
+with the objective of properly documenting it.
+
+To showcase Sphinx automatic documentation generation capabilities
+you will use Python, which is the default :term:`domain`
```
I think mentioning the default domain is not needed. It's not related to autodoc at all.

astrojuanlu replied:
Fair enough 👍🏽

tk0miya commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,198 @@
+
+.. code-block:: bash
+
+   $ python -m venv .venv
```
Our installation document does not require to install the venv library. So it might not be available in some Linux distro. For example, it is failed in Ubuntu 20.04:
```
root@b5f8ca310971:/# python3 -m venv .venv
The virtual environment was not created successfully because ensurepip is not
available.  On Debian/Ubuntu systems, you need to install the python3-venv
package using the following command.

    apt-get install python3-venv

You may need to use sudo with that command.  After installing the python3-venv
package, recreate your virtual environment.

Failing command: ['/.venv/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']
```

astrojuanlu replied:
You're right, this was a concern for me as well. This is a very Ubuntu and Debian-specific problem, but a very prevalent and relevant one. I'd rather keep using virtual environments though. Do you think we should expand `installation.rst` to account for this?

tk0miya replied:
IMO, it's important for beginners not familiar with python especially. It's needed to mention in installation.rst or here. It's not a "required" package, but a "recommended" one for the tutorial.

astrojuanlu replied:
I think it makes more sense in `installation.rst`. I'll update it on this PR if you don't mind.

tk0miya replied:
+1

astrojuanlu replied:
Done!

stevepiercy replied:
I would like to use `env` instead of `.venv` to disambiguate between the package `venv` and the virtual environment directory `.venv`. It's also a couple of characters shorter to type, it is visible in operating system UIs, and it's less of a mouthful to pronounce when read aloud. Trainers who would use this tutorial's material in presentations will thank you for it.

tk0miya commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,198 @@
+  as well as some extra configuration keys.
+- ``source/index.rst``: The :term:`master document` of the project,
+  which serves as welcome page
+  and contains the root of the "table of contents tree" (or *toctree*).
```
I think it's better to use a definition list instead. What do you think?

astrojuanlu replied:
Good idea! Will change it

tk0miya commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,198 @@
+---------------
+
+Setting up our project and development environment
+..................................................
```
We usually uses `~~~~` for the 3rd level headings. Could you use it?

astrojuanlu replied:
Sure, didn't notice that. Changing!

tk0miya commented on doc/contents.rst:
```diff
@@ -7,6 +7,7 @@ Sphinx documentation contents
    :maxdepth: 2
 
    usage/index
+   tutorial/index
```
If my understanding correct, this tutorial will replace the "Getting Started" page, right?
https://www.sphinx-doc.org/en/master/usage/quickstart.html

If so, this tree should be placed at the top of the document, I think.

Note: I understand this is still not matured. So I can agree to put this here until enough mature.

astrojuanlu replied:
Good point. Initially, I was not thinking about replacing the `quickstart.rst`, but it's true that, when it matures, it will be a better starting point.

I agree with moving it at the top, perhaps in a later pull request, when this has a bit more content?

tk0miya replied:
Okay, let's replace it with this after matured.

tk0miya reviewed (approved):
Thank you for your contribution!
LGTM with nits!

jakobandersen commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,198 @@
+that will serve as a guide throughout the process,
+with the objective of properly documenting it.
+
+To showcase Sphinx automatic documentation generation capabilities
```
Note, *automatic* documentation generation with standalone Sphinx is as far as I know only available for Python.
How about having first a section on the manual documentation using domain directives (which indeed translates quite similarly to other domains), and then an additional section with autodoc for Python?
Thereby the underlying concepts are introduced and then the shorthands on top of that.

astrojuanlu replied:
It's an excellent point @jakobandersen, I was a bit misguided here. Will rephrase it so it makes more sense.

astrojuanlu committed:
Consistent heading styles

astrojuanlu committed:
Style

astrojuanlu committed:
Use more appropriate terminology for emphasized text

astrojuanlu committed:
Amend code documentation introduction

astrojuanlu commented:
Thanks a lot for the prompt review @tk0miya , and for your comment @jakobandersen ! I addressed most of the comments. Happy to lay the first stone of this!

tk0miya commented:
Thank you for your update. LGTM again.
Do you have a plan to update this more? I need to know what is a good time for merging this.

I'll wait for a while for reviews from other contributors.

astrojuanlu commented:
> Do you have a plan to update this more?

Only the `installation.rst` part, that'll be quick. Apart from that, I consider this complete.

astrojuanlu committed:
Move virtual environment explanation to installation

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+Setting up our project and development environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+On a new directory,
```
```suggestion
In a new directory,
```

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+
+   (.venv) $ sphinx-quickstart docs
+
+This will present you a series of questions
```
```suggestion
This will present to you a series of questions
```

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+Creating the documentation layout
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Then, from the command line,
```
```suggestion
Then from the command line,
```

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+
+This will present you a series of questions
+required to create the basic directory and configuration layout for your project
+inside the `docs/` folder.
```
```suggestion
inside the ``docs`` folder.
```

Its name is "docs", not "docs/".

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+This will present you a series of questions
+required to create the basic directory and configuration layout for your project
+inside the `docs/` folder.
+To proceed, introduce these answers:
```
Answer each question as follows.

astrojuanlu replied:
I observed that you tend to remove colons at the end of sentences that end in "as follows", whereas you are okay with some other sentences that precede enumerations or instructions. I'd be keen to learn more, would you please share some explanations about it?

astrojuanlu replied:
English.SE seems to disagree https://english.stackexchange.com/a/394875/20057

stevepiercy replied:
First I should preface that the dominant practice should be continued, but I don't know what that is for Sphinx. A style guide should be adopted. @tk0miya is there an established preferred style in Sphinx for how to end a sentence when followed by a list or code block?

There are varying opinions about ending a sentence with a colon or a period when followed by a _list_, but not many about following it with _code_. The context and audience need to be considered. For example, technical documentation versus a narrative tutorial. Even American and English English don't agree. No one way is correct, as they are all opinions. I may have been inconsistent when spotting these instances and reviewing them. Apologies for that.

Some guides prefer to end all sentences with a period and never a colon. I prefer this style because it prevents the confusion of what punctuation to use when ending a sentence, as they _always_ end with a period without exception. It avoids sentence fragments, such as "For example:" and encourages the writer to form complete sentences. This is especially helpful for visually impaired readers and screen readers. It prevents the confusion of whether to use `:` or `::` when a code block follows to get syntax highlighting.

"As follows", "following", and similar phrases are hints to readers—especially visually impaired or people with reading comprehension impairments—that something is coming up related to this sentence, so I almost always include them.

tk0miya replied:
>is there an established preferred style in Sphinx for how to end a sentence when followed by a list or code block?

AFAIK, there are no policies and rules for them.

IMO, reST provides `::` syntax to notate a code-block after a paragraph
```
blah blah blah::

    here is a code block.
```

This is rendered as following.
```
blah blah blah:

    here is a code block.
```

So I think the sentences that end with a colon just before code-block are commonly used in reST documents. I've usually written them also. To imitate the colon at the tail of the sentence, I've usually use a single colon and `code-block` directive as follows

```
blah blah blah:

.. code-block:: some-highlighting language

   here is a code block.
```

On the other hand, I don't see the sentences that end with a colon just before bullet-list much.

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+- ``> Project release []``: Write "``0.1``" (without quotes)
+  and press :kbd:`Enter`.
+- ``> Project language [en]``: Leave it empty (the default, English)
+  and press :kbd:`Enter`.
```
Are the `>` actually displayed on screen? If not, then please omit them. When rendered to HTML, there are both bullet points and `>` symbols which is visually distracting.

astrojuanlu replied:
This is the output of `sphinx-quickstart`:

```
$ sphinx-quickstart docs
Welcome to the Sphinx 3.5.4 quickstart utility.

Please enter values for the following settings (just press Enter to
accept a default value, if one is given in brackets).

Selected root path: docs

You have two options for placing the build directory for Sphinx output.
Either, you use a directory "_build" within the root path, or you separate
"source" and "build" directories within the root path.
> Separate source and build directories (y/n) [n]: y

The project name will occur in several places in the built documentation.
> Project name: Lumache
> Author name(s): Graziella
> Project release []: 0.1

If the documents are to be written in a language other than English,
you can select a language here by its language code. Sphinx will then
translate text that it generates into that language.

For a list of supported codes, see
https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-language.
> Project language [en]: 

Creating file /tmp/docs/source/conf.py.
Creating file /tmp/docs/source/index.rst.
Creating file /tmp/docs/Makefile.
Creating file /tmp/docs/make.bat.

Finished: An initial directory structure has been created.

You should now populate your master file /tmp/docs/source/index.rst and create other documentation
source files. Use the Makefile to build the docs, like so:
   make builder
where "builder" is one of the supported builders, e.g. html, latex or linkcheck.
```

stevepiercy replied:
Thank you. This is fine.

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+It is written in reStructuredText,
+a powerful markup language.
+
+Modify the file like follows:
```
```suggestion
Modify the file as follows.
```

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+- an **inline external link**,
+- and a ``note`` **admonition**.
+
+Now, to render it with the new content,
```
```suggestion
Now to render it with the new content,
```

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+
+Now, to render it with the new content,
+you can use the ``sphinx-build`` command as before,
+or leverage the convenience script like this:
```
```suggestion
or leverage the convenience script as follows.
```

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+.. code-block:: bash
+
+   (.env) $ cd docs/
+   (.env) $ make html
```
```suggestion
   (.env) $ make html
```

Whichever name you end up using for the virtual environment, please be consistent.

astrojuanlu replied:
Good catch!

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+- two examples of **inline markup**: ``**strong emphasis**`` (typically bold)
+  and ``*emphasis*`` (typically italics),
+- an **inline external link**,
+- and a ``note`` **admonition**.
```
Suggest referencing each term to the Sphinx docs instead of adding strong emphasis.

```suggestion
- and a ``note`` :ref:`rst-directives`.
```

I considered a reference to docutils with [`:dudir:`](https://github.com/sphinx-doc/sphinx/blame/bfca913f1695969f8cb37bfde5ab8609b3b7a907/doc/usage/restructuredtext/basics.rst#L334), but I think that may lead the newcomer astray. Maybe not.

stevepiercy replied:
Alternatively, we could add explicit labels above each bullet point in the Sphinx docs to have a more specific reference than `rst-directives`.

astrojuanlu replied:
Changed 👍🏽

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+
+On a new directory,
+create a file called ``README.rst``
+with the following contents:
```
```suggestion
with the following content.
```

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+  and press :kbd:`Enter`.
+
+After the last question,
+you will see the new ``docs`` directory with some content::
```
Let's be explicit about the syntax highlighting. This tree diagram renders with the CSS class `highlight-default`, which is most likely Python, and therefore might render strangely depending on the theme.

```suggestion
you will see the new ``docs`` directory with the following content.

.. code-block:: text
```

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+		├── _static
+		└── _templates
+
+These files are:
```
These items are described as follows.

astrojuanlu replied:
To avoid repeating "as follows" every time, I changed it to something slightly different.

stevepiercy commented on doc/usage/installation.rst:
```diff
@@ -185,6 +185,30 @@ the ``--pre`` flag.
+
+When installing Sphinx using pip,
+it is highly recommended to use *virtual environments*,
+which isolate the Installation
```
```suggestion
which isolate the installed packages from the system packages,
```

stevepiercy commented on doc/usage/installation.rst:
```diff
@@ -185,6 +185,30 @@ the ``--pre`` flag.
+When installing Sphinx using pip,
+it is highly recommended to use *virtual environments*,
+which isolate the Installation
+and remove the need to use administrator privileges.
```
```suggestion
thus removing the need to use administrator privileges.
```

stevepiercy commented on doc/usage/installation.rst:
```diff
@@ -185,6 +185,30 @@ the ``--pre`` flag.
+
+::
+
+   $ python -m venv .venv
```
One more reason to use `env` instead of `.venv`, the PyPA uses this in its documentation:

`python3 -m venv env`

or

`py -m venv env`

Being consistent with the PyPA docs is a very Good Thing™.

astrojuanlu replied:
> Being consistent with the PyPA docs is a very Good Thing™.

I might be excessively nitpicky with this, but the Python Packaging Authority has stated several times that they are not an authority, despite the name :) Many people ([including myself](https://twitter.com/juanluisback/status/1398623726070022144)) have fallen this trap several times.

astrojuanlu replied:
Having said that:

1. All options seem to be somewhat used, judging at the output of https://www.gitignore.io/api/python:

```
# Environments
# .env
.env/
.venv/
env/
venv/
ENV/
env.bak/
venv.bak/
pythonenv*
```

2. Consistency with the PyPA docs is a good thing in itself, but [A Foolish Consistency is the Hobgoblin of Little Minds](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds). I like `.venv` because it's hidden, and I avoid `env` or `.env` to remove confusion with [direnv](https://direnv.net/) and similar tools. However, I'm sure they have their own reasons to adopt `env`, and probably either name is just fine.
3. On the other hand, giving it a different name from the PyPA docs also communicates that it doesn't really matter, or that the user can change it if they want.
4. Like all serious things in life, [let's decide them with a Twitter poll](https://twitter.com/juanluisback/status/1399375962228224005) (I'm joking, but anyway I'm curious about the results)

stevepiercy replied:
1. OK. I see `env` is common usage.
2. Would this consistency with the PyPA example be foolish?
3. That would be a sidebar to creating a virtual environment.
4. There's also strong opinions about where to put them, either inside the project folder or in the user home directory.

For tutorials, I strongly advocate for keeping things as simple as possible on all levels to avoid confusion. If you have ever conducted any training, you know how difficult it is to get people to install packages into a virtual environment in the first place. That's the background to my reasons for using `env` in my [original comment](https://github.com/sphinx-doc/sphinx/pull/9276#discussion_r641983256).

pradyunsg replied:
> `python3 -m venv env`

Only because I need to update https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/?highlight=venv#creating-a-virtual-environment. 😅 

In general, I think a lot of the material is converging toward `.venv` and the fact that packaging.python.org is lagging is due to lack of volunteer time to update it more than anything else.

> Being consistent with the PyPA docs is a very Good Thing™.

WELL. I guess it's a good thing that I'm currently working on improving PyPA documentation stuff then. ;)

pradyunsg replied:
> Python Packaging Authority has stated several times that they are not an authority, despite the name :)

As an aside, there is definitely authority associated with the PyPA now -- Python Steering Council officially delegates power for deciding on all Python Packaging things to the PyPA. I think we're aware of the "what we say matters" situation, but that is not as coordinated/structured as folks would expect from "authority". :)

pradyunsg replied:
All this is to say, `.venv` works and I don't think it needs to change to be in sync with packaging.python.org; although I'll likely update that to match this style later. :)

stevepiercy commented on doc/usage/installation.rst:
```diff
@@ -185,6 +185,30 @@ the ``--pre`` flag.
+
+.. warning::
+
+   Note like in some Linux distributions like Debian and Ubuntu
```
```suggestion
   Note that in some Linux distributions, such as Debian and Ubuntu,
```

stevepiercy commented on doc/usage/installation.rst:
```diff
@@ -185,6 +185,30 @@ the ``--pre`` flag.
+.. warning::
+
+   Note like in some Linux distributions like Debian and Ubuntu
+   this might require an extra installation step::
```
```suggestion
   this might require an extra installation step as follows.

    .. code-block:: bash
   
```

stevepiercy commented on doc/usage/installation.rst:
```diff
@@ -185,6 +185,30 @@ the ``--pre`` flag.
+   Note like in some Linux distributions like Debian and Ubuntu
+   this might require an extra installation step::
+
+     $ apt-get install python3-venv
```
Remove prompts as they are often copy-pastad and they are different across operating systems.
```suggestion
      apt-get install python3-venv
```

astrojuanlu replied:
I hear your point about the prompt being pasted from time to time, but

1. It's a good way of communicating that it's a Linux/UNIX-like command (unlike, say, Windows), and
2. It keeps consistency with other code blocks that retain the Bash output

What do you think?

stsewd replied:
I think you can use `.. code:: console` to avoid having those being copied.

astrojuanlu replied:
TIL: https://pygments.org/docs/lexers/?highlight=console#pygments.lexers.shell.BashSessionLexer

Thanks @stsewd ! Changing it everywhere in this PR now.

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+Getting started
+---------------
+
+Setting up our project and development environment
```
```suggestion
Setting up your project and development environment
```
and underline to go with this.

stevepiercy commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,213 @@
+After the last question,
+you will see the new ``docs`` directory with some content::
+
+	docs/
```
```suggestion
	docs
```
I noticed that `docs/`, `build`, and `source` are directories, but only `docs/` has a trailing slash. Suggest consistency by removing the trailing slash, as it is not actually included in its name.

astrojuanlu replied:
Interesting: I copy-pasted the result from the `tree` command, and the trailing slash is retained if the user specifies it:

```
(.venv) $ tree -L 1 docs
docs
├── build
├── make.bat
├── Makefile
└── source

2 directories, 2 files
(.venv) $ tree -L 1 docs/
docs/
├── build
├── make.bat
├── Makefile
└── source

2 directories, 2 files
```

I agree with keeping it consistent, will change.

stevepiercy reviewed (changes requested):
This is an excellent start. There are bunch of subtle things that people who write documentation will appreciate, including the use of the second person voice. That is difficult to pull off well, and I congratulate you.

I have a few points for your consideration.

1. Please do not use semantic line wrapping. There are strong advocates for this style, but it has several problems, including:
    - Nobody writes real prose like that.
    - It is alien to people who do not write English as their first language.
    - It hurts readability of the source, which ends up looking like lit-magzine prose-poetry.
    - It has no effect on rendered output, which is where the value should be.
    - Adds considerable difficulty to create translations.
    - It is inconsistent with the current arbitrarily and historically chosen 79-character line wrapping (which itself is also not helpful, IMO)

    If any new style should be adopted, I would strongly advocate for the simple to understand "one line per sentence" style.
    - Easy to remember.
    - Easy to translate.
    - Easy to read diffs.
    - No more rewrapping entire paragraphs when making a single typo correction.
    - Code editors have supported soft line wraps for decades.
2. Use 4 spaces for indenting `code-blocks`, list items, and other things. For an explanation, please see https://stackoverflow.com/questions/48311159/is-3-space-indentation-required-in-rest/48313531#48313531 and the "bonus tip".
3. Please leave this open to allow feedback. This is Memorial Day Weekend in the United States, and some folks extend a 3-day weekend into a 9-day holiday.

tk0miya commented:
>Please do not use semantic line wrapping. There are strong advocates for this style, but it has several problems, including:

I don't have a big opinion about this. But it's like a bikeshed discussion. I believe it must not be a blocker of improvement. So this issue is not good to discuss. How about below?

* Until the new rule is determined, follow the current rule (folded by 80 chars)
* Discuss the new rule in the new issue (if somebody would like to do. I'm not interested)

>Use 4 spaces for indenting code-blocks, list items, and other things. For an explanation, please see https://stackoverflow.com/questions/48311159/is-3-space-indentation-required-in-rest/48313531#48313531 and the "bonus tip".

This is also.

At present, almost of our document is rewritten with this (implicit) rule:

* 2 spaces for bullet lists
* 3 spaces for directive contents
* 4 spaces for other indentation

>Please leave this open to allow feedback. This is Memorial Day Weekend in the United States, and some folks extend a 3-day weekend into a 9-day holiday.

Okay, I'll keep this open until the end of the next weekend.

astrojuanlu commented:
Thanks for the super detailed feedback @stevepiercy , really appreciate it! Yes, I am not a native English speaker myself (as you might have guessed already) and I spent a lot of time getting used to using the second person in tutorials, I'm happy the expert eye spotted it :)

About semantic linefeeds, you raise valid criticism and I'd like to discuss it further. Therefore, as @tk0miya said, I'll rewrap all text in 80 columns and open an issue about style later on (after I'm done addressing your comments - to avoid messing everything up).

About indentation, I wanted to be consistent with the rest of Sphinx, although I might have made some mistakes. I prefer to stick to what @tk0miya said.

Going over your comments now.

astrojuanlu committed:
Style fixes

Co-authored-by: Steve Piercy <web@stevepiercy.com>

astrojuanlu committed:
Links to relevant reST syntax explanation

astrojuanlu committed:
Be consistent with directory names

astrojuanlu committed:
Be consistent with virtual environment name

astrojuanlu committed:
More style changes

astrojuanlu commented:
I think the `docslint` failure is unrelated to this PR.

astrojuanlu commented:
In case the GitHub UI hides them, I left a bunch of comments unresolved, all the rest were addressed. Thanks again!

astrojuanlu committed:
Change Pygments lexer to account for prompt

tk0miya commented:
> I think the `docslint` failure is unrelated to this PR.

I fixed them it in https://github.com/sphinx-doc/sphinx/pull/9303. Could you merge the HEAD of 4.x into this branch?

tk0miya commented:
>Okay, I'll keep this open until the end of the next weekend.

It seems all discussions are resolved (I know we still have a discussion for doc-styling). So I'd like to merge this if nobody objects within a day.

@stevepiercy Do you have comment for the updates? Please let us know if you find another point.

stevepiercy commented:
@tk0miya I'll defer style discussion to other issues for the sake of merging this PR. It makes sense to continue existing practices, and thank you for clarifying them where they exist.

I have only one objection, specifically about the name of the virtual environment. I am very strongly in favor of keeping it short, simple, and pronounceable because I would use this material to conduct trainings. Two fewer characters to type and syllables to speak, especially when repeated throughout the docs, would be relief to students and the instructor. Compare "vee-ee-en-vee space dot-vee-ee-en-vee" versus "vee-ee-en-vee space ee-en-vee". If it were just a single usage, I would not care, but it would be used throughout the material.

astrojuanlu commented:
> I fixed them it in #9303. Could you merge the HEAD of 4.x into this branch?

Sure! Do you usually do merge or rebase? (I prefer the latter, but I want to be consistent with what's usually done in Sphinx)

About the name of the virtual environment, my Twitter poll of 32 votes yielded a clear advantage over `.venv` https://mobile.twitter.com/juanluisback/status/1399375962228224005 I know it's by no means a big sample, but it matches my observations of (a) keeping it hidden and (b) avoiding overlap with direnv (Brett Cannon agrees https://mobile.twitter.com/brettsky/status/1399439424929230850)

astrojuanlu committed:
Merge branch '4.x' into new-tutorial

astrojuanlu commented:
(I went ahead and merged)

astrojuanlu committed:
Adjust line length

stevepiercy commented:
I have doubts a student of this tutorial uses `direnv`, but if they do, they are most likely not a Sphinx newbie, they know what they are doing with virtual environments, and they will have their own opinions and preferences for them.

I am advocating for the audience of Sphinx newbies and the trainers who will teach them using this material. I feel very strongly about this. I really don't want to spend time explaining to Python newbies the difference between the command "venv" and the name of a virtual environment that happens to be "dot-venv". It's hard enough explaining what is a Python virtual environment in the first place.

astrojuanlu commented:
I don't have anything meaningful to add here, already voiced my technical arguments. I am also a Python instructor and I don't see much difficulty, but this is subjective. Don't want to die on this particular hill and I am not a fan of bikeshedding, so I defer to @tk0miya to make the final call, and I'll be happy to make changes if needed.

tk0miya commented:
I don't have any opinion for the name of virtualenv directory.

In my experience, it's difficult to teach what virtualenv is for non python developers. So, personally, it's too much to use it on the tutorials. I know, my opinion is a bit extreme. But the installation of python itself and its packages is too complicate to describe (conda, pyenv, virtualenv, etc. etc. etc.). So I'd not like to touch the topic...

tk0miya commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,216 @@
+
+To showcase Sphinx capabilities for code documentation
+you will use Python,
+which also supports *automatic* documentation generation.
```
In my understanding, texts should be folded by 80 chars instead of semantic line folding until the discussion, right? It seems some updates are needed.

astrojuanlu replied:
Sorry, a misunderstanding from my side. Fixing.

astrojuanlu replied:
Done!

astrojuanlu committed:
Rewrap text

pradyunsg reviewed (approved):
This feels like a great first iteration. I think this can be merged as is now! :)

pradyunsg commented:
> So, personally, it's too much to use it on the tutorials. I know, my opinion is a bit extreme. But the installation of python itself and its packages is too complicate to describe (conda, pyenv, virtualenv, etc. etc. etc.). So I'd not like to touch the topic...

*puts on documentation author hat*
*puts on pip maintainer hat*
*puts on past maintainer of virtualenv hat*

And... as you've probably guessed, I disagree that we should avoid it mentioning virtual environments here. 

The goal of a tutorial is NOT to provide the user with _all_ their options, but rather to give them a step-by-step guided learning journey. I think we should _definitely_ recommend using virtual environments, especially in tutorial style content. One of the big reasons to introduce them early is to protect beginners from the problems of messing up their OS, especially on Linux. Avoiding mentioning virtual enviroments risks breakage due to inexperienced users breaking their operating systems.

FWIW, I am planning on making a round of improvements to packaging.python.org at some point in the near future to make it easier to defer virtualenv-specific stuff to there, but I think what this tutorial is doing right now, is basically perfect -- not going into too much detail, but still doing the right thing.

astrojuanlu commented:
@tk0miya Any chance to get this merged? In any case, I already started working on the second part, and I hope to send another PR this week.

tk0miya commented:
@pradyunsg Thank you for the comment. Okay, let's go with `.venv`.  Merging now.

tk0miya commented:
Merged. @astrojuanlu Good work!

And, thank you for all :-)

astrojuanlu commented:
Thanks to you! On to the next part 💪🏽

jfbu commented:
I encounter a
```
! Package inputenc Error: Unicode character ˈ (U+02C8)
(inputenc)                not set up for use with LaTeX.
```
when building the sphinx.pdf after this. I wll investigate...

jfbu commented:
@astrojuanlu I am not familiar with phonetics. Would it be ok to replace `/luˈmake/` with `/lu'make/` ?
Else I can add to our conf.py for latex pdf build a `\DeclareUnicodeCharacter{02C8}{\textquotesingle}` but perhaps there is a better way involving loading a special font for characters used in phonetics?

jfbu commented:
Is this rewrap intended:

```rest
.. code-block:: text

    docs ├── build ├── make.bat ├── Makefile └── source ├── conf.py ├──
    index.rst ├── _static └── _templates
```

? I think it crept in at some point unintended...

astrojuanlu commented:
Oh right, that was unintended... I thought I had fixed it locally, my bad. Thanks @jfbu 

I am not an expert either, but I don't think this warrants an extra unicode character, probably your suggested change is good enough. Will fix both things soon.

astrojuanlu commented:
#9339

jfbu commented on doc/tutorial/index.rst:
```diff
@@ -0,0 +1,185 @@
+directory and configuration layout for your project inside the ``docs`` folder.
+To proceed, answer each question as follows:
+
+- ``> Separate source and build directories (y/n) [n]``: Write "``y``" (without
+  quotes) and press :kbd:`Enter`.  - ``> Project name``: Write "``Lumache``"
+  (without quotes) and press :kbd:`Enter`.  - ``> Author name(s)``: Write
+  "``Graziella``" (without quotes) and press :kbd:`Enter`.  - ``> Project
+  release []``: Write "``0.1``" (without quotes) and press :kbd:`Enter`.  - ``>
+  Project language [en]``: Leave it empty (the default, English) and press
+  :kbd:`Enter`.
```
This paragraph looks as if it has been reflowed and was initially a list.  I find strange the inline use of `-` separators.

astrojuanlu replied:
You're right, thanks a lot. Not having a brilliant day today! Will add it to #9339
